src/ipaperftest/core/main.py

```diff
@@ -0,0 +1,438 @@
+        for conf in self.generate_clients(ctx):
+            if conf:
+                self.machine_configs.append(conf)
+
+        # Related: https://github.com/hashicorp/vagrant/issues/4967
+        vagrant_additional_config = (
+            "config.ssh.insert_key = false\n"
+            'config.ssh.private_key_path = ["~/.vagrant.d/insecure_private_key", "%s"]'
+            % ctx.params['private_key']
+            if ctx.params.get('private_key')
+            else ""
+        )
+
+        file_contents = VAGRANTFILE_TEMPLATE.format(
+            vagrant_additional_config=vagrant_additional_config,
+            machine_configs="\n".join(self.machine_configs),
+        )
+        with open("Vagrantfile", "w") as f:
+            f.write(file_contents)
+
+    def create_vms(self, ctx):
+        """Bring up all the configured virtual machines"""
+        print("Creating VMs...")
+        sp.run(["vagrant", "up", "--parallel"])
+
+    def collect_hosts(self, ctx):
+        """Collect IP information on the configured hosts in a dict
+
+           hostname:ip
+        """
+        grep_output = sp.run(
+            "vagrant ssh-config | grep -i HostName -B 1",
+            shell=True, stdout=sp.PIPE).stdout.decode("utf-8")
+        host_lines = grep_output.split("--")
+        for line in host_lines:
+            pair = line.replace("\n", "").replace("Host ", "").split("  HostName ")
+            self.hosts[pair[0]] = pair[1]
+
+    def generate_ansibile_inventory(self, ctx):
+        # Each replica should have main server + all the other replicas
+        # as servers so the topology is built correctly
+        replica_lines = []
+        for name, _ in self.hosts.items():
+            if name.startswith("client") or name.startswith("server"):
+                continue
+            servers = ["server." + self.domain]
+            for other, _ in self.hosts.items():
+                if other.startswith("replica") and other != name:
+                    servers.append(other + "." + self.domain)
+            replica_lines.append(name + " ipareplica_servers={}".format(",".join(servers)))
+
+        inventory_str = HOSTS_FILE_TEMPLATE.format(
+            domain=self.domain.lower(),
+            realm=self.domain.upper(),
+            client_hostnames="\n".join(
+                [name for name, _ in self.hosts.items() if name.startswith("client")]
+            ),
+            replica_hostnames="\n".join(replica_lines),
+        )
+        with open("hosts", "w") as f:
+            f.write(inventory_str)
+
+    def generate_ssh_config(self, ctx):
+        sp.run("vagrant ssh-config > vagrant-ssh-config",
+               shell=True, stdout=sp.PIPE)
+
+    def ansible_ping(self, ctx):
+        print("Sending ping to VMs...")
+        sp.run(
+            "ansible -i hosts all -m ping --ssh-extra-args '-F vagrant-ssh-config'",
+            shell=True,
+        )
+
+        # TODO: raise an exception instead?
```

Comment:
I left this as a warning because even if not all clients are provisioned, the run might still be useful (for example, you request 150 but only get 147). Anyhow, since we fixed the networking issues this is quite difficult to happen.